### PR TITLE
Overall timeout policy feature

### PR DESCRIPTION
**Pull request by NYMEZIDE**
HttpClientSettings: add new optional parameter TimeoutOverall
When HttpClientTimeout less than TimeoutOverall then HttpClientTimeout equal as TimeoutOverall.

tests: HttpClientWrapperBuilder using AddJsonClient instead of a direct call AddHttpClient
add 5 test cases of Overall Timeout parameter

Issue #13

### CODE REVIEW

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:11]: Do we really have to make it nullable? It adds unnecessary complexity to the code and to the usage. Actually the reasonable defaults looks better I think, because the main goal of the library is to simplicity.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:9]: The most arguable thing here is should we really support both HttpClientTimeout and TimeoutOverall? Yes, they are techincally different things, but for the end-user prospective it looks like it could be the same thing. My http request will be completed or failed after N seconds. The reason in both cases - timeout. Maybe we should work with single timeout setting? What do you think?

File: src/Dodo.HttpClient.ResiliencePolicies.Tests/TimeoutPolicyTests.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies.Tests/TimeoutPolicyTests.cs:70]: Your tests are very good. Great job!

**Review by Ceridan — CHANGES_REQUESTED** [review]
Great job actually, this issue get off the ground. I have some questions to discuss before approval. See review.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:44]: It is a bit tricky thing, but it is possible to do. Actually HttpClientSettings already have information about selected retry strategy. So, it can calculate the sum of all retry attempts. It will be not exact value for strategy with jitter backoffs, but good enough to validate overall timeout.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:11]: Maybe it is good idea to leave **only** TimeoutOverall and TimeoutPerTry? For the HttpClient we can calculate Timeout like this: `Timeout = TimeoutOverall + delta`, where `delta` is some hard-coded constant time. We need it just because some policies like Fallback or Caching may sit between TimeoutPolicy and Client itself.
Yes it is a breaking change, but we can do it because it is a library and we can change major version for such changes. It also allows us to get rid of nulls and make our code much cleaner.

Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:11]: I was thinking about preservation backward compatibility.
So I decided to both introduce a new parameter and make it optional.

The idea of unifying the parameter looks good.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:9]: I agree and will prepare the change.

File: src/Dodo.HttpClient.ResiliencePolicies.Tests/TimeoutPolicyTests.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies.Tests/TimeoutPolicyTests.cs:70]: Thank you, it's a pleasure.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:11]: You may not think about backward compatibility. Just change PR base from `master` to `milestone/v2.0.0`. We will include all breaking changes to new major version of the library.

File: src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs:7]: Currently we can remove this const. It is not in use anymore.

File: src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs:9]: Could I ask you to make some minor changes:
1. Let's set TimeoutOverall less than 100000 miliseconds. For example 50000 miliseconds by default, because the default HttpClient timeout equals 100000 miliseconds. If someone will create it's own HttpClient (not by AddJsonClient method provided by the library) and apply our default policies, than we will have both TimeoutOverall and HttpClient timeout equal to the same value and the result could be unpredictable.
2. I want to keep the same order of constants and parameters as it was before. I mean that TimeoutOverall is a replacement for HttpClientTimeout, I want to preserve this order. Just move this line before the TimeoutPerTry. It would be easier for our users to understand the changes in the version 2.0: replacement of one thing with another thing, all other stuff is the same.
3. You have typo in the name TimeoutOverallInMillicesons. Typo in the word Milliseconds.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:10]: Preserve properties order: Move TimeoutOverall before TimeoutPerTry

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:17]: Preserver order in the constructors: Move TimeoutOverall as first argument (because HttpClientTimeout was the first),

File: src/Dodo.HttpClient.ResiliencePolicies.Tests/DSL/HttpClientWrapperBuilder.cs
Comment (Ceridan) [inline src/Dodo.HttpClient.ResiliencePolicies.Tests/DSL/HttpClientWrapperBuilder.cs:15]: _uri could be marked as readonly. Also we could just inline this.

**Review by Ceridan — CHANGES_REQUESTED** [review]
Great job, currently we have good consistent code with proper OverallTimeout integration. Could I ask you to make some minor changes to make this changes easier to understand for end-users.

File: src/Dodo.HttpClient.ResiliencePolicies.Tests/DSL/HttpClientWrapperBuilder.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies.Tests/DSL/HttpClientWrapperBuilder.cs]: I fix it.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs]: Agree. Done.

File: src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies/HttpClientSettings.cs:9]: Done.

File: src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs
Comment (NYMEZIDE) [reply src/Dodo.HttpClient.ResiliencePolicies/Defaults.cs]: All done.

**Review by Ceridan — APPROVED** [review]
Awesome job! Thank you a lot!

**Pull request merged**
